### PR TITLE
Add parsing for more COS event types, to Machine/ContainerState

### PR DESCRIPTION
**Pull request by jkl73**
Add parsing for more COS event types, converting them to Machine/ContainerState

Add parsing for ImageId, EnvVars and Args

### CODE REVIEW

File: cel/cos_tlv.go [hunk 1]
```diff
@@ -79,3 +85,38 @@ func (t TLV) ParseToCosTlv() (CosTlv, error) {
...
+	if !utf8.ValidString(value) {
+		return fmt.Errorf("malformed env value: [%s] (contains non-utf8 character)", value)
+	}
+	match, err := regexp.MatchString("^[a-zA-Z_][a-zA-Z0-9_]*$", name)
+	if err != nil {
+		// only if the pattern is invalid
+		panic(err)
+	}
+	if !match {
+		return fmt.Errorf("malformed env name: [%s] (doesn't match the regex ^[a-zA-Z_][a-zA-Z0-9_]*$)", name)
```
Comment (alexmwu) [inline cel/cos_tlv.go:104]: should we have this be more human reaadable? something like "env name must start with an alpha character or _ , followed by a string of alphanumeric characters or _ "

**Review by alexmwu — APPROVED** [review]

File: cel/cos_tlv.go [hunk 2]
```diff
@@ -79,3 +85,38 @@ func (t TLV) ParseToCosTlv() (CosTlv, error) {
...
+	}
+	return nil
+}
+
+// ParseEnvVar takes in environment variable as a string (foo=bar), parses it and return its name
+// and value, or an error if it fails the validation check.
+func ParseEnvVar(envvar string) (string, string, error) {
+	e := strings.SplitN(string(envvar), "=", 2)
+	if len(e) < 2 {
```
[the remark this answers is not in the document]
Comment (jkl73) [reply cel/cos_tlv.go:113]: Done

File: cel/cos_tlv.go (hunk 1 above)
Comment (jkl73) [reply cel/cos_tlv.go:104]: added the description

File: cel/cos_tlv_test.go [hunk 3]
```diff
@@ -60,3 +68,48 @@ func TestCosEventlog(t *testing.T) {
...
+		envname     string
+		envvalue    string
+		expectedErr error
+	}{
+		{"normal case 1", "foo=bar", "foo", "bar", nil},
+		{"normal case 2", "FOO=1", "FOO", "1", nil},
+		{"normal case 3", "SESSION_MANAGER=\"`\\local/:@?%/tmp/.u/1,unix/.com:/tmp/.u/5\"", "SESSION_MANAGER", "\"`\\local/:@?%/tmp/.u/1,unix/.com:/tmp/.u/5\"", nil},
+		{"empty value", "foo=", "foo", "", nil},
+		{"multiple =", "foo=bar=baz=", "foo", "bar=baz=", nil},
+		{"bad name", "3foo=bar=baz=", "", "", fmt.Errorf("malformed env name: [%s] (doesn't match the regex ^[a-zA-Z_][a-zA-Z0-9_]*$)", "3foo")},
```
[the remark this answers is not in the document]
Comment (jkl73) [reply cel/cos_tlv_test.go:85]: I change the expectedErr to a string, let me know what you think

File: cel/cos_tlv.go [hunk 4]
```diff
@@ -79,3 +85,38 @@ func (t TLV) ParseToCosTlv() (CosTlv, error) {
 func (t TLV) IsCosTlv() bool {
 	return t.Type == CosEventType
 }
+
+// ValidateEnvVar takes in a environment name and its value, run some checks and return an error
+// if they are invalid.
+func ValidateEnvVar(name string, value string) error {
```
Comment (josephlr) [inline cel/cos_tlv.go:91]: Should we have this be called something like `FormatEnvVar` that also does the joining with `=`?

That way `FormatEnvVar` and `ParseEnvVar` would be inverses of eachother and any caller of `ValidateEnvVar` will always be formatting them as the very next call.

**Review by josephlr — APPROVED** [review]

File: cel/cos_tlv_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (josephlr) [reply cel/cos_tlv_test.go]: This seems fine, we can later switch to using structured errors if this approach runs into scaling or ergonomic issues.

File: cel/cos_tlv.go [hunk 5]
```diff
@@ -79,3 +85,38 @@ func (t TLV) ParseToCosTlv() (CosTlv, error) {
...
+		return fmt.Errorf("malformed env name, env name must start with an alpha character or '_', followed by a string of alphanumeric characters or '_' (^[a-zA-Z_][a-zA-Z0-9_]*$): [%s]", name)
+	}
+	return nil
+}
+
+// ParseEnvVar takes in environment variable as a string (foo=bar), parses it and returns its name
+// and value, or an error if it fails the validation check.
+func ParseEnvVar(envvar string) (string, string, error) {
+	e := strings.SplitN(string(envvar), "=", 2)
+	if len(e) < 2 {
```
[the remark this answers is not in the document]
Comment (jkl73) [reply cel/cos_tlv.go:113]: done

File: cel/cos_tlv.go [hunk 6]
```diff
@@ -79,3 +85,38 @@ func (t TLV) ParseToCosTlv() (CosTlv, error) {
...
+		return fmt.Errorf("malformed env name, contains non-utf8 character: [%s]", name)
+	}
+	if !utf8.ValidString(value) {
+		return fmt.Errorf("malformed env value, contains non-utf8 character: [%s]", value)
+	}
+	match, err := regexp.MatchString("^[a-zA-Z_][a-zA-Z0-9_]*$", name)
+	if err != nil {
+		// only if the pattern is invalid
+		panic(err)
+	}
```
[the remark this answers is not in the document]
Comment (jkl73) [reply cel/cos_tlv.go:102]: Done

File: cel/cos_tlv.go (hunk 4 above)
Comment (jkl73) [reply cel/cos_tlv.go:91]: yes that makes sense

File: cel/cos_tlv.go [hunk 7]
```diff
@@ -79,3 +85,38 @@ func (t TLV) ParseToCosTlv() (CosTlv, error) {
...
+	if !match {
+		return fmt.Errorf("malformed env name, env name must start with an alpha character or '_', followed by a string of alphanumeric characters or '_' (^[a-zA-Z_][a-zA-Z0-9_]*$): [%s]", name)
```
[the remark this answers is not in the document]
Comment (jkl73) [reply cel/cos_tlv.go:104]: Done

**Pull request merged**